src/types/Currency.sol

```diff
@@ -30,43 +30,77 @@ library CurrencyLibrary {
-                        or(eq(mload(0x00), 1), iszero(returndatasize())), // Returned 1 or nothing.
-                        call(gas(), currency, 0, 0x10, 0x44, 0x00, 0x20)
+                // Get a pointer to some free memory.
+                let fmp := mload(0x40)
+
+                // Write the abi-encoded calldata into memory, beginning with the function selector.
+                mstore(fmp, 0xa9059cbb00000000000000000000000000000000000000000000000000000000)
+                mstore(add(fmp, 4), and(to, 0xffffffffffffffffffffffffffffffffffffffff)) // Append and mask the "to" argument.
+                mstore(add(fmp, 36), amount) // Append the "amount" argument. Masking not required as it's a full 32 byte type.
+
+                success :=
+                    and(
+                        // Set success to whether the call reverted, if not we check it either
+                        // returned exactly 1 (can't just be non-zero data), or had no return data.
+                        or(and(eq(mload(0), 1), gt(returndatasize(), 31)), iszero(returndatasize())),
+                        // We use 68 because the length of our calldata totals up like so: 4 + 32 * 2.
+                        // We use 0 and 32 to copy up to 32 bytes of return data into the scratch space.
+                        // Counterintuitively, this call must be positioned second to the or() call in the
+                        // surrounding and() call or else returndatasize() will be zero during the computation.
+                        call(gas(), currency, 0, fmp, 68, 0, 32)
                     )
-                ) {
-                    mstore(0x00, 0xf27f64e4) // `ERC20TransferFailed()`.
-                    revert(0x1c, 0x04)
-                }
-                mstore(0x34, 0) // Restore the part of the free memory pointer that was overwritten.
+
+                // clean the memory we used
+                mstore(fmp, 0) // 4 byte `selector` and 28 bytes of `to` were stored here
+                mstore(add(fmp, 0x20), 0) // 4 bytes of `to` and 28 bytes of `amount` were stored here
+                mstore(add(fmp, 0x40), 0) // 4 bytes of `amount` were stored here
+            }
+            if (!success) {
+                selector = ERC20TransferFailed.selector;
+            }
+        }
+
+        // revert containing the bubbled up error as an argument
+        assembly ("memory-safe") {
+            if iszero(success) {
+                let size := returndatasize()
+                let fmp := mload(0x40)
+
+                // Encode selector, offset, size, data
+                mstore(fmp, selector)
+                mstore(add(fmp, 0x04), 0x20)
+                mstore(add(fmp, 0x24), size)
+                returndatacopy(add(fmp, 0x44), 0, size)
+
+                // Ensure the size is a multiple of 32 bytes
+                let encodedSize := add(0x44, mul(div(add(size, 31), 32), 32))
+                revert(fmp, encodedSize)
             }
```

Comment:
Line 91 ~ 103 , used in src/libraries/Hooks.sol.
Suggest we can extract it and reuse it as a public lib.
But now it is fine.

Reply:
cool, added a issue to track this. will update as soon as i am available

https://github.com/pancakeswap/pancake-v4-core/issues/121